### PR TITLE
Dkg scenario tests

### PR DESCRIPTION
**Pull request by wrinkledeth**
Test suite created for DKG Scenario testing (commitDkg and postProccessDkg).

- dkgHelper() created to streamline commitDkg testing.
- setPhase() created to set phase during ppDkg testing.
- Multiple commitDkg and ppDkg scenario tests written
(Controller code coverage went from 80 -> 95%)!

Scenario details and notes can be found in DKGScenarioTests.md

### CODE REVIEW

File: contracts/test/DKGScenarioTests.t.sol [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+        params[0] = Params(node1, false, err, 0, 3, publicKey, partialPublicKey1, new address[](0));
+        params[1] = Params(node2, false, err, 0, 3, publicKey, partialPublicKey2, new address[](0));
+        params[2] = Params(node3, false, err, 0, 3, publicKey, partialPublicKey3, disqualifiedNodes);
+        params[3] = Params(node4, false, err, 0, 3, publicKey, partialPublicKey4, disqualifiedNodes);
+        params[4] = Params(node5, false, err, 0, 3, publicKey, partialPublicKey5, disqualifiedNodes);
+        dkgHelper(params);
+        // printGroupInfo(0);
+
+        // * Is this okay?
+        // * When non-disqualified majority members < g.threshold (3), group is not formed, nodes are not slashed.
```
Comment (kafeikui) [inline contracts/test/DKGScenarioTests.t.sol:344]: This is actually an edge case in commitDkg, since a bad node tends not to report itself contained in the  disqualifiedNodes. However I think it's ok as honest nodes will trigger postProcessDkg later for getting arranged, where bad nodes in disqualifiedNodes will get slashed. Also, if the size of majority members with disqualified nodes excluded is less than threshold, the group should not get formed.

**Review by kafeikui — APPROVED** [review]
LGTM. Good job! We made a great progress to have 95% code coverage. 

Next, some cases still need to be covered. The first is other scenarios triggering dkg process. For now all DKG scenarios are triggered by node register, while they can also be triggered by node exit and rebalance happening in node register/exit. Also, to cover rebalance cases, we have to do some setup work to set the context of network. 

e.g. we have group_0 which has 5 members working, then 1 member of group_0 wants to exit the network, then controller will let 4 members left in group_0 do dkg as 4 > 3 which is the threshold i.e. the group still meet the grouping condition, after that, in happy path group_0 will able to work again.

e.g. we have group_0 which has 5 members working, and group_1 which has 3 members working, then 1 member of group_1 wants to exit the network, then controller will let group_1 which has 2 members left rebalance with group_0, in happy path after the two dkgs group_0(3 members) and group_1(4 members) will both able to work.

In above cases we need to pay attention to pathes containing rebalance, and please assert group.epoch in each case. Since you have tested different cases in a single dkg process, I think we can cover only happy path in above scenarios, or the better, all the pathes we have for now.

Second, we need to test the strategy of grouping. For now we have a new-started network and 5 fresh nodes to register, but what if we already have some nodes and groups in the network and some node wants to register and some node wants to leave? We need to assert controller's behavior of grouping decision, as to let what nodes to do the next round of dkg as group x, epoch n+1.

(in the following cases we set groupMaxCapacity as 6 instead of 10 to make it simpler to setup, you can change it in controller setting)

e.g. we have group_0 and group_1 having 6 members each, denoted as (6,6), and here comes a new node, then the controller should make grouping decision as creating a new group_2 and adding the new node into group_2 and making rebalance between group_0 and group_2, as a result the network status should be (3,6,4), in happy path group_0 and group_2 will both able to work.

e.g. we have group_0 and group_1 having 3 members each, denoted as (3,3), and they both able to work, and here comes a new node, then the controller should make grouping decision as creating a new group_2 and adding the new node into group_2 and trying to rebalance, as a result the network status should be (3,3,1), in happy path group_2 won't able to work for now but have to stay for the next two nodes to register.

e.g. we have group 0-4 having 3 members each, and they all able to work, (3,3,3,3,3), and here comes a new node, then the controller should make grouping decision as adding the new node into group_0, as a result the network status should be (4,3,3,3,3), in happy path group_0 will able to work again.

e.g. we have group 0-4 having 6 members each, and they all able to work, (6,6,6,6,6), and here comes a new node, then the controller should make grouping decision as creating a new group_5 and adding the new node into group_5 and making rebalance between group_0 and group_5, as a result the network status should be (3,6,6,6,6,4), in happy path group_0 will able to work again. This may looks similar to the first case, but please note that the paths to initial status (3,6) and (6,6,6,6,6) are different and they are both reachable in real cases. 

e.g. we have group_0 and group_1 in the network, (6,3), and here a member of group_1 wants to quit, then the controller should make grouping decision as making rebalance between group_0 and group_2, as a result the network status should be (4,4), in happy path group_0 and group_1 will both able to work again.

e.g. we have group_0 and group_1 in the network, (6,3), and here a member of group_0 wants to quit, then the controller should make grouping decision as emitting dkg event of group_0 with 5 members left, as a result the network status should be (5,3), in happy path group_0 will able to work again.

I know it is a hard work, but I think it's more proper to have them covered in forge testing environment instead of integration test involving controller contract, user contract and nodes. Please tell me if there is anything I can help.

I think we can merge this pr first. Then I will incorporate staking contract into the current repo. It may take some changes into the current codebase but we can mock the staking contract while we test the controller to keep the test cases easy to go. Let's keep moving simultaneously.

**Comment by wrinkledeth** [comment]
Thanks for the detailed response. I understand what needs to be done and I'll start working on those additional testcases as well as improving the framework to support such testing.
